File: src-tauri/src/commands/profiles.rs
//! Profile and step commands: validate, delegate, return.
//!
//! F5 adds the file door and the review gate to this surface: a profile can be
//! written out as text, read back in as text, imported **unreviewed**, and then
//! accepted a step at a time. The two file commands are given a path the person
//! chose in the system's own dialog — this side never picks one, and never
//! looks anywhere it was not sent.

use std::path::PathBuf;

use tauri::State;

use crate::db::models::{ImportStep, Profile, Step};
use crate::db::{profiles, Db};
use crate::error::Result;
use crate::os::files;

#[tauri::command]
pub fn profiles_list(db: State<'_, Db>) -> Result<Vec<Profile>> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::list_profiles(&conn)
}

#[tauri::command]
pub fn profile_create(db: State<'_, Db>, name: String) -> Result<Profile> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::create_profile(&conn, &name)
}

#[tauri::command]
pub fn profile_rename(db: State<'_, Db>, id: String, name: String) -> Result<Profile> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::rename_profile(&conn, &id, &name)
}

#[tauri::command]
pub fn profile_delete(db: State<'_, Db>, id: String) -> Result<()> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::delete_profile(&conn, &id)
}

#[tauri::command]
pub fn steps_list(db: State<'_, Db>, profile_id: String) -> Result<Vec<Step>> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::list_steps(&conn, &profile_id)
}

#[tauri::command]
pub fn step_add(
    db: State<'_, Db>,
    profile_id: String,
    kind: String,
    config_json: String,
    timing_json: Option<String>,
    wait_json: Option<String>,
    place_json: Option<String>,
) -> Result<Step> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::add_step(
        &conn,
        &profile_id,
        &kind,
        &config_json,
        timing_json.as_deref().unwrap_or("{}"),
        wait_json.as_deref().unwrap_or("{}"),
        place_json.as_deref().unwrap_or("{}"),
    )
}

#[tauri::command]
pub fn step_update(
    db: State<'_, Db>,
    id: String,
    config_json: String,
    timing_json: Option<String>,
    wait_json: Option<String>,
    place_json: Option<String>,
) -> Result<Step> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::update_step(
        &conn,
        &id,
        &config_json,
        timing_json.as_deref().unwrap_or("{}"),
        wait_json.as_deref().unwrap_or("{}"),
        place_json.as_deref().unwrap_or("{}"),
    )
}

#[tauri::command]
pub fn step_delete(db: State<'_, Db>, id: String) -> Result<()> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::delete_step(&conn, &id)
}

/// Move a step one place up (`-1`) or down (`1`) within its profile.
#[tauri::command]
pub fn step_move(db: State<'_, Db>, id: String, direction: i64) -> Result<Vec<Step>> {
    let mut conn = db.0.lock().expect("the database lock was poisoned");
    profiles::move_step(&mut conn, &id, direction)
}

/// Store a profile that came from a file. It is stored **unreviewed**: the host
/// refuses to run it until every step has been accepted (ADR-013).
#[tauri::command]
pub fn profile_import(db: State<'_, Db>, name: String, steps: Vec<ImportStep>) -> Result<Profile> {
    let mut conn = db.0.lock().expect("the database lock was poisoned");
    profiles::import_profile(&mut conn, &name, &steps)
}

/// Accept one step of an imported profile.
#[tauri::command]
pub fn step_accept(db: State<'_, Db>, id: String) -> Result<Profile> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::accept_step(&conn, &id)
}

/// Accept every step of an imported profile at once.
#[tauri::command]
pub fn profile_accept(db: State<'_, Db>, id: String) -> Result<Profile> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    profiles::accept_profile(&conn, &id)
}

/// Read a profile file as text. What it means is the domain's business.
#[tauri::command]
pub fn profile_file_read(path: String) -> Result<String> {
    files::read_text(&PathBuf::from(path))
}

/// Write a profile file. The path is the one the person chose.
#[tauri::command]
pub fn profile_file_write(path: String, contents: String) -> Result<()> {
    files::write_text(&PathBuf::from(path), &contents)
}

File: src/data/profiles.ts
/**
 * The typed client for the profile and step commands.
 *
 * The host speaks snake_case, because that is what the schema speaks. The
 * interface speaks camelCase. The translation happens exactly once, here,
 * rather than leaking a database naming convention into every component.
 */

import { invoke } from '@tauri-apps/api/core';

import type { StepImport } from '@/domain/portable';
import {
  parseStoredStep,
  serializeStepConfig,
  type Problem,
  type Step,
  type StepConfig,
} from '@/domain/profile';
import { parseStoredWaitFor, serializeWaitFor, type WaitFor } from '@/domain/readiness';
import { parseStoredTiming, serializeTiming, type Timing } from '@/domain/timing';

export interface Profile {
  id: string;
  name: string;
  position: number;
  importedUnreviewed: boolean;
  createdAt: string;
  updatedAt: string;
}

/** A stored step, read back: either its configuration or why it could not be read. */
export type StoredStep =
  | { readable: true; step: Step }
  | {
      readable: false;
      id: string;
      profileId: string;
      position: number;
      reviewed: boolean;
      problems: Problem[];
    };

interface RawProfile {
  id: string;
  name: string;
  position: number;
  imported_unreviewed: boolean;
  created_at: string;
  updated_at: string;
}

interface RawStep {
  id: string;
  profile_id: string;
  position: number;
  kind: string;
  config_json: string;
  timing_json: string;
  wait_json: string;
  reviewed: boolean;
  created_at: string;
  updated_at: string;
}

function toProfile(raw: RawProfile): Profile {
  return {
    id: raw.id,
    name: raw.name,
    position: raw.position,
    importedUnreviewed: raw.imported_unreviewed,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

/**
 * A row that cannot be read is shown as such rather than dropped: a profile
 * with a step the interface silently hides is a profile that runs something
 * the person cannot see.
 */
function toStoredStep(raw: RawStep): StoredStep {
  const parsed = parseStoredStep(raw.kind, raw.config_json);
  const timing = parseStoredTiming(raw.timing_json);
  const waitFor = parseStoredWaitFor(raw.wait_json ?? '{}');
  if (!parsed.ok || Array.isArray(timing) || Array.isArray(waitFor)) {
    return {
      readable: false,
      id: raw.id,
      profileId: raw.profile_id,
      position: raw.position,
      reviewed: raw.reviewed,
      problems: [
        ...(parsed.ok ? [] : parsed.problems),
        ...(Array.isArray(timing) ? timing : []),
        ...(Array.isArray(waitFor) ? waitFor : []),
      ],
    };
  }
  return {
    readable: true,
    step: {
      id: raw.id,
      profileId: raw.profile_id,
      position: raw.position,
      config: parsed.config,
      timing,
      waitFor,
      reviewed: raw.reviewed,
    },
  };
}

export async function listProfiles(): Promise<Profile[]> {
  const raw = await invoke<RawProfile[]>('profiles_list');
  return raw.map(toProfile);
}

export async function createProfile(name: string): Promise<Profile> {
  return toProfile(await invoke<RawProfile>('profile_create', { name }));
}

export async function renameProfile(id: string, name: string): Promise<Profile> {
  return toProfile(await invoke<RawProfile>('profile_rename', { id, name }));
}

export async function deleteProfile(id: string): Promise<void> {
  await invoke('profile_delete', { id });
}

export async function listSteps(profileId: string): Promise<StoredStep[]> {
  const raw = await invoke<RawStep[]>('steps_list', { profileId });
  return raw.map(toStoredStep);
}

export async function addStep(
  profileId: string,
  config: StepConfig,
  timing: Timing,
  waitFor: WaitFor | null,
): Promise<StoredStep> {
  const raw = await invoke<RawStep>('step_add', {
    profileId,
    kind: config.kind,
    configJson: serializeStepConfig(config),
    timingJson: serializeTiming(timing),
    waitJson: serializeWaitFor(waitFor),
  });
  return toStoredStep(raw);
}

export async function updateStep(
  id: string,
  config: StepConfig,
  timing: Timing,
  waitFor: WaitFor | null,
): Promise<StoredStep> {
  const raw = await invoke<RawStep>('step_update', {
    id,
    configJson: serializeStepConfig(config),
    timingJson: serializeTiming(timing),
    waitJson: serializeWaitFor(waitFor),
  });
  return toStoredStep(raw);
}

export async function deleteStep(id: string): Promise<void> {
  await invoke('step_delete', { id });
}

/** Move a step one place up (`-1`) or down (`1`); the whole list comes back. */
export async function moveStep(id: string, direction: -1 | 1): Promise<StoredStep[]> {
  const raw = await invoke<RawStep[]>('step_move', { id, direction });
  return raw.map(toStoredStep);
}

/**
 * Store a profile that came from a file. It arrives unreviewed: the host
 * refuses to run it until every step has been accepted (ADR-013).
 */
export async function importProfile(name: string, steps: StepImport[]): Promise<Profile> {
  return toProfile(await invoke<RawProfile>('profile_import', { name, steps }));
}

/** Accept one step. The profile comes back, so the screen learns when the gate lifts. */
export async function acceptStep(id: string): Promise<Profile> {
  return toProfile(await invoke<RawProfile>('step_accept', { id }));
}

/** Accept every step of a profile at once. */
export async function acceptProfile(id: string): Promise<Profile> {
  return toProfile(await invoke<RawProfile>('profile_accept', { id }));
}

/** Read one file the person chose, as text. What it means is the domain's business. */
export async function readProfileFile(path: string): Promise<string> {
  return invoke<string>('profile_file_read', { path });
}

/** Write one file the person chose. */
export async function writeProfileFile(path: string, contents: string): Promise<void> {
  await invoke('profile_file_write', { path, contents });
}
